### DISCUSSION TITLE
Re-structure saving of programs

### DISCUSSION THREAD

**Discussion by TiBiBa** [hedyorg/hedy#1896] [Implementation discussions]
I noticed to be doing a lot of work on small fixes on the process of saving. For example, we now return a warning when a user tries to over-write an existing program. But, looking at the process of saving I noticed some other issues that do not seem so easy to fix with the current structure. Such as when continue working on a program, we simply "over-write" the program, not keeping track of the shared or submitted status. It might be worth the effort to take a look at re-writing the complete pipeline of saving.

Saving is currently cumbersome and it doesn't work intuitive for me (and I think neither for kids). As we already make small progress in #1885 to keep track of erroneous programs, it would be nice to implement a feature such as auto-saving with automatic parsing. First ideas that come to mind:

- Automatically store programs on x seconds interval
- Still enable users to manually save programs
- Naming with adventures is cumbersome and (to me) is not clear what we do
   -> We store the adventure name as a separate value but it is also in the program name?
- Keep track of changes in a program -> return a users the last x versions of their program
- When over-writing an existing name differentiate between new program and updating old one
- User's are unable to over-write a submitted program, they have to delete the old one first

But honestly, I have no clue how other online programming websites / tools such as Scratch implemented a saving environment. I will have a look at differentiate platforms to get some inspiration, but for now I would love some input with good ideas!

**Comment by StephanEggermont** [comment]
Is the amount of storage needed to provide (nearly)  full undo prohibitive? And keep explictly-saved versions?

**Comment by Felienne** [comment]
We have auto save now since #4057
